cni/pkg/nodeagent/informers_test.go

```diff
@@ -0,0 +1,516 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nodeagent
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/mock"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	"istio.io/istio/cni/pkg/util"
+	"istio.io/istio/pkg/config/constants"
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/monitoring/monitortest"
+	"istio.io/istio/pkg/test/util/assert"
+)
+
+func TestExistingPodAddedWhenNsLabeled(t *testing.T) {
+	setupLogging()
+	NodeName = "testnode"
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test",
+			Namespace: "test",
+		},
+		Spec: corev1.PodSpec{
+			NodeName: NodeName,
+		},
+		Status: corev1.PodStatus{
+			PodIP: "11.1.1.12",
+		},
+	}
+	ns := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "test"}}
+
+	client := kube.NewFakeClient(ns, pod)
+
+	// We are expecting at most 1 calls to the mock, wait for them
+	wg, waitForMockCalls := NewWaitForNCalls(t, 1)
+	fs := &fakeServer{testWG: wg}
+
+	fs.On("AddPodToMesh",
+		ctx,
+		pod,
+		util.GetPodIPsIfPresent(pod),
+		"",
+	).Return(nil)
+
+	server := &meshDataplane{
+		kubeClient: client.Kube(),
+		netServer:  fs,
+	}
+
+	server.Start(ctx)
+	handlers := setupHandlers(ctx, client, server, "istio-system")
+	client.RunAndWait(ctx.Done())
+	go handlers.Start()
+
+	// label the namespace
+	labelsPatch := []byte(fmt.Sprintf(`{"metadata":{"labels":{"%s":"%s"}}}`,
+		constants.DataplaneMode, constants.DataplaneModeAmbient))
+	_, err := client.Kube().CoreV1().Namespaces().Patch(ctx, ns.Name,
+		types.MergePatchType, labelsPatch, metav1.PatchOptions{})
+	assert.NoError(t, err)
+
+	waitForMockCalls()
+
+	assertPodAnnotated(t, client, pod)
+
+	// Assert expected calls actually made
+	fs.AssertExpectations(t)
+}
+
+func TestExistingPodAddedWhenDualStack(t *testing.T) {
+	setupLogging()
+	NodeName = "testnode"
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test",
+			Namespace: "test",
+		},
+		Spec: corev1.PodSpec{
+			NodeName: NodeName,
+		},
+		Status: corev1.PodStatus{
+			PodIPs: []corev1.PodIP{
+				{
+					IP: "11.1.1.12",
+				},
+			},
+		},
+	}
+	ns := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "test"}}
+
+	client := kube.NewFakeClient(ns, pod)
+
+	// We are expecting at most 1 calls to the mock, wait for them
+	wg, waitForMockCalls := NewWaitForNCalls(t, 1)
+
+	fs := &fakeServer{testWG: wg}
+
+	fs.On("AddPodToMesh",
```

Comment:
I won't block on this but I very strongly recommend against this pattern. It is not testing this way.

Why do we care some functions were called?? We should assert the end behavior -- the annotations were added.

https://matija-kovacek.medium.com/test-behaviour-not-implementation-d76af17603d0 etc (tbh I didn't read the article, just the title)

Reply:
The pod reconciliation logic in the watcher has many subtle conditionals - even in the old implementation. It frequently was broken subtly by PRs, and the minimal tests we did have would continue to pass because the "end result" (pod annotation) happened but other critical side effects did not, and were not verified, due to a lack of actual branch coverage.

It's a behaviorally critical block, and testing that the pod was annotated is not enough to actually verify the other side effects actually happened. Given the fragility and specificity of the code, a test that is similarly fragile and specific is a very good, very strong signal to guard against subtle breakage that results in pods being annotated but not actually having traffic captured. If you are intentionally breaking the logic, you should have no trouble intentionally fixing the test, and that will be visible in a PR.

These tests are a response to those problems and frequent breakages, and effectively prevent them - branch coverage is an imperfect but very hard to cheat metric for seeing if your tests _actually_ cover all behaviors, and in our internal impl we had very high branch coverage, which helped keep the baseline functionality extremely stable.

I'm happy to drop the mock library in a followup, but only if we can do so in a way that does not reduce branch code coverage unit test rates from what they currently are for this and other critical sections.

Also of course I think there's a lot we could do to rework that overly-complex function itself and make it less subtle, but again, happy to do that in followups, with tests, that prove no behavioral breakage.

Reply:
Yes, and in the previous code it was possible to change the code, and not the test, and break the behavior, but not the test - purely by accident, because the side effects were not tested.

That is not possible here, without, as you say, changing the test in two places. Which means CI can catch what it did not before.

As mentioned, happy to redo this and the underlying informer code more broadly in any way that does not reduce the branch coverage rates this solution gives us as a followup, but this code is harder to unintentionally break than it currently is in `master`, which is the immediate value.